### ISSUE TITLE
Change Record to be an interface

### ISSUE THREAD

**Issue by sa2aie** [surrealdb/surrealdb.net#88]
Change the class Record to be an interface instead of a class. This will be a more versatile solution.
This may be a breaking change due to a name change (Record->IRecord) to follow naming conventions.

**Comment by Odonno** [comment]
Indeed, having a base class and not an interface have many limitations. Migrating to an `IRecord` interface seems feasible. I was thinking about this implementation:

```csharp
public interface IRecord<T>
{
  public T? Id { get; set; }
}

public interface IThingRecord : IRecord<Thing> { }

public interface ITableAndIdRecord<TTable, TId> : IRecord<TId>
{
  public TTable Table { get; set; }
}
```

`IThingRecord` and `ITableAndIdRecord` as the 2 main base types to detect what to do: either use the `Thing` object directly or create a thing from both `Table` and `Id` property.

This will allow to create default interfaces and extend with new ones. Example:

```csharp
public interface IStringTableAndRecord<TId> : ITableAndIdRecord<string, TId> {}

public interface IStringTableAndStringIdRecord : IStringTableAndRecord<string> {}
```

I may have oversee things but that could fit your needs.

**Comment by uherman** [comment]
@Odonno This would be really great!

@sa2aie The breaking change could be avoided by inheriting the interface in the base class i belive.
```csharp
[Obsolete("Use IRecord interface instead")]
public abstract class Record : IRecord<Thing>
{
    public Thing Id { get; set; }
}

```
